// openfsm-warehouse-service-client-handler/WarehouseServiceClientHandler.js
// Коннектор  взаимодействия с WarehouseService (склад))
const fetch = require('node-fetch');
require('dotenv').config();


class WarehouseServiceClientHandler {
    constructor() {
        this.warehouseServiceGetProductIdUrl = process.env.WAREHOUSE_SERVICE_GET_PRODUCT_ID_URL;
        this.warehouseServiceGetProductsUrl = process.env.WAREHOUSE_SERVICE_GET_PRODUCTS_URL;
    }

    /**
     * Метод для выполнения запроса на получение параметров продукта
     * @param {string} product_id - идентификатор продукта
     * @returns {Object} - Объект с параметрами продукта
     */

    async getProductById(product_id) {
        try {
        // const queryParams = new URLSearchParams(obj).toString();
        // const response = await fetch(`${url}?${queryParams}`, {
            const response = await fetch(`${url}/${product_id}`, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            const data = await response.json();
            if (response.ok) {
                console.log(`User ${email} successfully logged in.`);
                return { success: true, data };
            } else {
                console.log(`Login failed for user ${email}: ${JSON.stringify(data)}`);
                return { success: false, status: response.status, data };
            }
        } catch (error) {
            console.log(`Error during login for user ${email}: ${error.message}`);
            return { success: false, error: error.message };
        }
    }

    async getProducts() {
        try {
            const response = await fetch(this.this.WAREHOUSE_SERVICE_GET_PRODUCTS_URL, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            const data = await response.json();
            if (response.ok) {
                console.log(`User ${email} successfully logged in.`);
                return { success: true, data };
            } else {
                console.log(`Login failed for user ${email}: ${JSON.stringify(data)}`);
                return { success: false, status: response.status, data };
            }
        } catch (error) {
            console.log(`Error during login for user ${email}: ${error.message}`);
            return { success: false, error: error.message };
        }
    }
}

module.exports = WarehouseServiceClientHandler;
