// openfsm-response-helper/ResponseHelper.js
const logger = require("./../../controllers/LoggerHandler"); // Работа с лог-файлами
const CommonFunctionHelper = require("openfsm-common-functions")


// Оберточный класс для Response

class ResponseHelper {
  constructor() {
  }

  setCode(code=null){
    this.code = code || 409;
    return this;
  }

  setStatus(status=null){
    this.status = status;
    return this;
  }

  setMessage(message=null){
    this.message = message;
    return this;
  }
  setData(data =null){
    this.data = data;
    return this;
  }
  
  getCode(){
    return this.code    
  }

  getStatus(){
    return this.status
  }

  getMessage(){
    return this.message 
  }

  getData(){
    return this.data;
  }
  
  destroy(){
      this.setCode()
      this.setStatus()
      this.setMessage()
      this.setData()
  }

  send(res, o = null) {    
    if(o) {
      this.setCode(o.code || 500)
      this.setStatus(o.status || false)
      this.setMessage(o.message || this.commonFunctionHelper.getDescriptionByCode(500))
      this.setData(o.data || null)
    }    
    if (this.getCode() >= 400) 
      logger.error(this.getMessage());

      let status  = this.getStatus() 
      let message = this.getMessage() 
      let data    = this.getData() 
      console.log({ status,  message, data  })
      res.status(this.getCode()).json((data) ? data : { status, message });
      this.destroy();
  }
}

module.exports = ResponseHelper
