// openfsm-response-helper/ResponseHelper.js
const logger = require("./../../controllers/LoggerHandler"); // Работа с лог-файлами
const CommonFunctionHelper = require("openfsm-common-functions")


// Оберточный класс для Response

class ResponseHelper {
  constructor() {
  }

  setCode(code){
    this.code = code;
    return this;
  }

  setStatus(status= undefined){
    this.status = status;
    return this;
  }

  setMessage(message = undefined){
    this.message = message;
    return this;
  }
  setData(data = undefined){
    this.data = data;
    return this;
  }
  
  getCode(){
    return this.code    
  }

  getStatus(){
    return this.status
  }

  getMessage(){
    return this.message 
  }

  getData(){
    return this.data || undefined;
  }
  
  send(res, o = null) {
    if(o) {
      this.setCode(o?.code || 500)
      this.setStatus(o?.status || false)
      this.setMessage(o?.message || this.commonFunctionHelper.getDescriptionByCode(500))
      this.setData(o?.data || undefined)
    }
    if (this.getCode() >= 400) logger.error(this.getMessage());
    let status = this.getStatus() 
    let message = this.getMessage() 
    let data = this.getData() 
    console.log({ status,  message, data  })
    res.status(this.getCode()).json((data) ? data : { status, message });
  }
}

module.exports = ResponseHelper
