// openfsm-order-dto/OrderDto.js
class OrderDto {

    constructor(dbRecord) {
        this.orderId = dbRecord.order_id || undefined;
        this.userId = dbRecord.user_id || undefined;
        this.price = dbRecord.price || undefined;
        this.status = dbRecord.status || 'NEW';
        this.createdAt = dbRecord.created_at || null;
        this.updatedAt = dbRecord.updated_at || undefined;
        this.transactionId = undefined;
    }


// Функция для проверки валидности статуса
     isValidOrderStatus(status) {
          return Object.values(OrderStatus).includes(status);
    }
/*
console.log(isValidOrderStatus('NEW')); // true
console.log(isValidOrderStatus('INVALID')); // false
*/
    // Getters
    getOrderId() {
        return this.orderId;
    }

    getUserId() {
        return this.userId;
    }

    getPrice() {
        return this.price;
    }

    getStatus() {
        return this.status;
    }

    getCreatedAt() {
        return this.createdAt;
    }

    getUpdatedAt() {
        return this.updatedAt;
    }

    getBillingTransactionId() {
        return this.billingTransactionId;
    }

    // Setters
    setOrderId(orderId) {
        this.orderId = orderId;
    }

    setUserId(userId) {
        this.userId = userId;
    }

    setPrice(price) {
        this.price = price;
    }

    setStatus(status) {
        if (Object.values(OrderStatus).includes(status)) {
            this.status = status;
        } else {
            throw new Error('Invalid status value');
        }
    }
    setCreatedAt(createdAt) {
        this.createdAt = createdAt;
    }

    setUpdatedAt(updatedAt) {
        this.updatedAt = updatedAt;
    }

    setBillingTransactionId(transactionId) {
        this.transactionId = transactionId;
    }
}

module.exports = OrderDto;
